src/constraint.rs

```diff
@@ -0,0 +1,399 @@
+                xor_constraints.push(constraints);
+            }
+        }
+
+        // do literal and global variable constraints first
+        // as they are the most "informative"
+        let literal_constraints = get_literal_and_global_constraints(&self.args, type_info);
+
+        match xor_constraints.len() {
+            0 => Err(TypeError::UnboundFunction(self.head)),
+            1 => Ok(literal_constraints
+                .chain(xor_constraints.pop().unwrap().into_iter())
+                .collect()),
+            _ => Ok(literal_constraints
+                .chain(once(Constraint::Xor(
+                    xor_constraints.into_iter().map(Constraint::And).collect(),
+                )))
+                .collect()),
+        }
+    }
+
+    fn to_expr(&self) -> Expr {
+        Expr::Call(
+            self.head,
+            self.args.iter().map(|arg| arg.to_expr()).collect(),
+        )
+    }
+}
+
+fn get_literal_and_global_constraints<'a>(
+    args: &'a [AtomTerm],
+    type_info: &'a TypeInfo,
+) -> impl Iterator<Item = Constraint<AtomTerm, ArcSort>> + 'a {
+    args.iter().filter_map(|arg| {
+        match arg {
+            AtomTerm::Var(_) => None,
+            // Literal to type constraint
+            AtomTerm::Literal(lit) => {
+                let typ = type_info.infer_literal(lit);
+                Some(Constraint::Assign(arg.clone(), typ.clone()))
+            }
+            AtomTerm::Global(v) => {
+                if let Some(typ) = type_info.global_types.get(v) {
+                    Some(Constraint::Assign(arg.clone(), typ.clone()))
+                } else {
+                    panic!("All global variables should be bound before type checking")
+                }
+            }
+        }
+    })
+}
+
+/// Construct a set of `Assign` constraints that fully constrain the type of arguments
+pub fn simple_constraints(
+    name: Symbol,
+    arguments: &[AtomTerm],
+    sorts: &[ArcSort],
+) -> Vec<Constraint<AtomTerm, ArcSort>> {
+    if arguments.len() != sorts.len() {
+        vec![Constraint::Impossible(
+            ImpossibleConstraint::ArityMismatch {
+                atom: Atom {
+                    head: name,
+                    args: arguments.to_vec(),
+                },
+                expected: sorts.len(),
+                actual: arguments.len(),
+            },
+        )]
+    } else {
+        arguments
+            .iter()
+            .cloned()
+            .zip(sorts.iter().cloned())
+            .map(|(arg, sort)| Constraint::Assign(arg, sort))
+            .collect()
+    }
+}
+
+/// Construct a set of `Eq` constraints for the given arguments.
+/// If a sort is given, all arguments should also be equivalent to that sort.
+/// If a length is given, the arguments should have the exact length.
+/// If an output is given, the last element of arguments should have this output.
+/// useful for polymorphic and var-arg functions.
+/// Requires arguments.len() > 0
+pub fn all_equal_constraints(
```

Comment:
This helper function is a bit confusing because it does a lot at once.
What if we had a "contraint builder" system? It would have functions like `with_exact_length` and `with_all_arguments_type` and `with_output_type`.

Reply:
That's a great idea! I've made the requested refactoring, which also makes the interface look nicer.